hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManager.java

```diff
@@ -512,25 +523,60 @@ private void handleUnderReplicatedContainer(final ContainerInfo container,
       if (source.size() > 0) {
         final int replicationFactor = container
             .getReplicationFactor().getNumber();
-        final int delta = replicationFactor - getReplicaCount(id, replicas);
+        // Want to check if the container is mis-replicated after considering
+        // inflight add and delete.
+        // Create a new list from source (healthy replicas minus pending delete)
+        List<DatanodeDetails> targetReplicas = new ArrayList<>(source);
+        // Then add any pending additions
+        targetReplicas.addAll(replicationInFlight);
+
+        int delta = replicationFactor - getReplicaCount(id, replicas);
+        final ContainerPlacementStatus placementStatus =
+            containerPlacement.validateContainerPlacement(
+                targetReplicas, replicationFactor);
+        final int misRepDelta = placementStatus.additionalReplicaRequired();
+        final int replicasNeeded
+            = delta < misRepDelta ? misRepDelta : delta;
+
         final List<DatanodeDetails> excludeList = replicas.stream()
             .map(ContainerReplica::getDatanodeDetails)
             .collect(Collectors.toList());
+        excludeList.addAll(replicationInFlight);
```

Comment:
Do we add the in-flight replications twice here this way, or I am missing something?

Reply:
Well spotted. I think I can remove the logic after this line, as I already have created that list earlier in the method now:

```
        if (actionList != null) {
          actionList.stream().map(r -> r.datanode)
              .forEach(excludeList::add);
        }
```

Reply:
👍